### ISSUE TITLE
Cross Translation Unit (CTU) Static Analysis of Curl with Clang

### ISSUE THREAD

**Issue by martong** [curl/curl#4589]
Dear Curl Community,

[Clang's CTU static analysis](https://clang.llvm.org/docs/analyzer/user-docs/CrossTranslationUnit.html) has evolved a lot in this year. We have chosen Curl
as one of the projects which we continuously analyze in our CI loop to
safeguard the quality of CTU analysis. We thought that the results might be
interesting for you Curl developers too. So, we stored the [results for one run on a 
demo server for version 7.66.0](http://codechecker-demo.eastus.cloudapp.azure.com/Default/#run=Curl_curl-7_66_0_CTU&review-status=Unreviewed&review-status=Confirmed&detection-status=New&detection-status=Reopened&detection-status=Unresolved&tab=Curl_curl-7_66_0_CTU). Please use demo/demo to log in. If you find
any false positive or if you have any other comments then we would be happy to receive
your feedback.

Sincerely,
Gabor Marton (CTU developer)

**Comment by martong** [comment]
@dkrupp @gyorb

**Comment by bagder** [comment]
Thanks a lot for helping us improve the project!

(I don't think you should include the test code in the report - anything under `tests/`, or if you do, mark them separately since I don't think many of us will consider warnings in that area to be terribly important.)

Then about warnings in the product code, let's start with the warning at llist.c:73
`Access to field 'prev' results in a dereference of a null pointer (loaded from field 'head')`

I can't see how this is true. From where I sit, the `head` pointer is only NULL if there are no entries in the list and then that code path isn't taken. Can you see how it happens?

I took a look at the warning at strcase.c:111 which says it is a possible dereference of a NULL pointer the first argument, from the tool_getparam.c:1561:

https://github.com/curl/curl/blob/curl-7_66_0/src/tool_getparam.c#L1561

... but I fail to see how this can be true, because `nextarg` is not allowed to be NULL when the option is of type `ARG_STRING` (or higher) which this argument is. From the code here:

https://github.com/curl/curl/blob/9cd755e1d768bbf228e7c9faf223b7459f7e0105/src/tool_getparam.c#L587-L596

I'll continue through the rest of the warnings too...

**Comment by martong** [comment]
Thank you for your feedback.

> but I fail to see how this can be true, because nextarg is not allowed to be NULL when the option is of type ARG_STRING (or higher) which this argument is.

My understanding is that the analyzer is `assuming field 'desc' is < ARG_STRING` and this way `nextarg = (char *)&parse[1];` is not executed symbolically.

Anyway, any false positive finding can help us to further improve the analyzer.
Also, usually an assert may guide (e.g. `assert(nextarg)`) the analyzer to avoid a false assumption.

**Comment by bagder** [comment]
The run listed 72 "reports". I selected "only show unique" (to hide different code paths hitting the same issue) and it said 22.

Out of those 22, 16 concerned code in `tests/` which aren't part of any shipped product and therefore not terribly important and I ignore those.

Out of the remaining 6 reports I found only false positives (two mentioned above already) and warnings due to conditionally compiled code.

I found no report to act on.

Bonus feedback: a site *without HTTPS* offered in 2019? That's just crazy, especially since these scanner results can very well end up very sensitive!
